Unicode support Part 1

MrPerson opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Makes all calls to ascii2text() and text2ascii() unicode aware as well as all code in code/__HELPERS/text.dm.

Adds defines for the footext_char() procs to maintain 512 support.

I did some cleanup on the text helpers since a lot of them were really, really bad.

reject_bad_text() and reject_bad_name() have an arg to reject non-ascii chars which defaults to TRUE.


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Gonna need support for those eggplant emojis... and I guess Cyrillic 

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MrPerson committed:
Unicode support Part 1

Makes all calls to ascii2text() and text2ascii() unicode aware as well as all calls in code/__HELPERS/text.dm.

Adds defines for the footext_char() procs to maintain 512 support.

I did some cleanup on the text helpers since a lot of them were really, really bad.

reject_bad_text() and reject_bad_name() have an arg to reject non-ascii chars which defaults to TRUE.

RaveRadbury commented:
Is this going to fix things like "John Doe&#39;s Awesome Paystand"?

MrPerson commented:
Probably not, that kind of thing is down to html encoding and decoding, not reading actual unicode.

MrPerson committed:
Give travis a more recent beta

MrPerson commented:
I don't know what to do about the failed check, seems above my pay grade.

TheVekter commented:
> I don't know what to do about the failed check, seems above my pay grade.

You're getting undefined proc errors when it's compiled.

SpaceManiac commented:
argh, now I'm going to have to rebuke all these uses of the `_char` functions

MrPerson committed:
Ultimate compat
Added the _char procs I was missing
Test the build instead of the version because Lummox is a goddamn moron

AnturK commented on code/__HELPERS/text.dm:
```diff
@@ -109,80 +115,98 @@
-	if(last_char_group == 1)
-		t_out = copytext(t_out,1,length(t_out))	//removes the last character (in this case a space)
+	if(last_char_group == SPACES_DETECTED)
+		t_out = copytext_char(t_out, 1, -1) //removes the last character (in this case a space)
```
If you know it's space you don't really need _char right ?

MrPerson replied:
http://jkorpela.fi/chars/spaces.html

There are spaces besides " " char 32.

AnturK replied:
But you don't check for these.

MrPerson replied:
Well yeah but if someone ever does want to, now they don't have to change this line or, more likely, forget to.

AnturK reviewed (commented):
Time to get on encoding pain train.

SpaceManiac commented:
This needs to be thoroughly reviewed, I'm not at all sold on the `_char` proc family

MrPerson commented:
Sold on it working correctly or sold on it being needed at all? Like I'm gonna be up front here: I literally didn't know anything about UTF-8 a week ago, so you're probably right.

MrPerson committed:
Suggested fixes, removes the last of the dd_ helpers

SpaceManiac commented on code/__HELPERS/text.dm:
```diff
@@ -40,8 +40,8 @@
-			t = copytext(t, 1, index) + repl_chars[char] + copytext(t, index+1)
-			index = findtext(t, char, index+1)
+			t = copytext(t, 1, index) + repl_chars[char] + copytext(t, index + length(char))
+			index = findtext(t, char, index + length(char))
```
Correct. :+1:

SpaceManiac reviewed (changes requested):
I did not finish reviewing `code/__HELPERS/text.dm` but this should give you some idea of why I disapprove of the `_char` functions

MrPerson commented:
I'm getting mixed signals on if we should favor the byte-based or char-based procs. I'm trying to split the difference and use byte-based counts when looping over the entire string or when it was easy and char-based ones when necessary.

Listen we can clean this up later, I'm just trying to get things functional, that's all.

MrPerson commented on code/game/objects/items/devices/PDA/cart.dm:
```diff
@@ -308,8 +308,8 @@ Code:
 					for(var/obj/machinery/power/apc/A in L)
-						menu += copytext(add_tspace(A.area.name, 30), 1, 30)
-						menu += " [S[A.equipment+1]] [S[A.lighting+1]] [S[A.environ+1]] [add_lspace(DisplayPower(A.lastused_total), 6)]  [A.cell ? "[add_lspace(round(A.cell.percent()), 3)]% [chg[A.charging+1]]" : "  N/C"]<BR>"
+						menu += copytext_char(add_trailing(A.area.name, 30, " "), 1, 30)
```
Outside the scope for this tbh. If you want I can just get rid of the copytext outright but either way I'm not getting into this kind of stuff right now.

SpaceManiac replied:
Fine to punt this out of scope, but I wanted to point out that `_char` does not help

MrPerson replied:
You know what, I'll toss in a comment saying as such

MrPerson commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -391,9 +391,13 @@
+				var/ascii = 0
+				for(var/i=1, i<=len, i += length(char))
+					char = string[i]
+					ascii = text2ascii(char)
```
I would still have to call i += length(string[i]) so there's no improvement there

SpaceManiac commented:
> I'm getting mixed signals

If it's another maintainer here let me know and I'll hash it out with them.

I respect the effort but I'd rather we do this right than rely on cheap stopgaps.

MrPerson committed:
Reading the reference is very important!

MrPerson commented:
At this point everyone would just defer to you anyway since you clearly know what you're talking about.

If by "cheap stopgags" you mean "just replace all text procs with _char versions everywhere", I agree that's dumb and I would love to not call _char procs at all. I'll try to think of ways to get rid of more of them.

ninjanomnom commented on code/_compile_options.dm:
```diff
@@ -39,6 +39,26 @@
 #endif
 
+//Compatability -- These procs were added in 513.1493, not 513.1490
+//Which really shoulda bumped us up to 514 right then and there but instead Lummox is a dumb dumb
```
Rapid compatibility breakages is exactly what beta releases are for

MrPerson committed:
Minor code improvements and a comment

MrPerson commented:
That twitterize() proc's use of copytext_char() could probably be rewritten using regex but what's here is functional. The rest of the length_char() calls are kind of wrong for the same reason the PDA-area name use case is wrong, but there's not much to do about that now.

optimumtact commented:
the reality is that most of the text handling SHOULD be replaced by regex use, rather than trying to treat strings like arrays of characters

Where that is not possible we should use the _char forms for pragmatism

Cyberboss commented on code/_compile_options.dm:
```diff
@@ -39,6 +39,26 @@
+#define findtext_char(args...) findtext(args)
+#define findtextEx_char(args...) findtextEx(args)
+#define findlasttext_char(args...) findlasttext(args)
+#define findlasttextEx_char(cargs...) findlasttextEx(args)
```
Typo

MrPerson committed:
Typo
